File: app/src/main/java/com/ywwynm/everythingdone/model/HabitRecord.java
package com.ywwynm.everythingdone.model;

import android.database.Cursor;
import android.support.annotation.IntDef;

import org.joda.time.DateTime;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by ywwynm on 2016/2/11.
 * model layer for table "habit_records"
 */
public class HabitRecord {

    public static final int TYPE_FINISHED = 0;
    public static final int TYPE_CANCEL_FINISHED = 1;
    public static final int TYPE_FAKE_FINISHED = 2;
    public static final int TYPE_FAKE_CANCEL_FINISHED = 3;

    @IntDef(value = {
            TYPE_FINISHED,
            TYPE_CANCEL_FINISHED,
            TYPE_FAKE_FINISHED,
            TYPE_FAKE_CANCEL_FINISHED
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface Type {}

    private long id;
    private long habitId;
    private long habitReminderId;
    private long recordTime;
    private int recordYear;
    private int recordMonth;
    private int recordWeek;
    private int recordDay;
    private @Type int type;

    public HabitRecord(long habitId, long habitReminderId) {
        this.id = 0;
        this.habitId = habitId;
        this.habitReminderId = habitReminderId;
        this.recordTime = System.currentTimeMillis();
        DateTime dt = new DateTime(recordTime);
        this.recordYear = dt.getYear();
        this.recordMonth = dt.getMonthOfYear();
        this.recordWeek = dt.getWeekOfWeekyear();
        this.recordDay = dt.getDayOfMonth();
        this.type = TYPE_FINISHED;
    }

    public HabitRecord(long habitId, long habitReminderId, long recordTime) {
        this.id = 0;
        this.habitId = habitId;
        this.habitReminderId = habitReminderId;
        this.recordTime = recordTime;
        DateTime dt = new DateTime(recordTime);
        this.recordYear = dt.getYear();
        this.recordMonth = dt.getMonthOfYear();
        this.recordWeek = dt.getWeekOfWeekyear();
        this.recordDay = dt.getDayOfMonth();
        this.type = TYPE_FINISHED;
    }

    public HabitRecord(long id, long habitId, long habitReminderId, long recordTime,
                       int recordYear, int recordMonth, int recordWeek, int recordDay, @Type int type) {
        this.id = id;
        this.habitId = habitId;
        this.habitReminderId = habitReminderId;
        this.recordTime = recordTime;
        this.recordYear = recordYear;
        this.recordMonth = recordMonth;
        this.recordWeek = recordWeek;
        this.recordDay = recordDay;
        this.type = type;
    }

    public HabitRecord(Cursor c) {
        this(c.getLong(0), c.getLong(1), c.getLong(2), c.getLong(3),
                c.getInt(4), c.getInt(5), c.getInt(6), c.getInt(7), c.getInt(8));
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getHabitId() {
        return habitId;
    }

    public void setHabitId(long habitId) {
        this.habitId = habitId;
    }

    public long getHabitReminderId() {
        return habitReminderId;
    }

    public void setHabitReminderId(long habitReminderId) {
        this.habitReminderId = habitReminderId;
    }

    public long getRecordTime() {
        return recordTime;
    }

    public void setRecordTime(long recordTime) {
        this.recordTime = recordTime;
    }

    public int getRecordYear() {
        return recordYear;
    }

    public void setRecordYear(int recordYear) {
        this.recordYear = recordYear;
    }

    public int getRecordMonth() {
        return recordMonth;
    }

    public void setRecordMonth(int recordMonth) {
        this.recordMonth = recordMonth;
    }

    public int getRecordWeek() {
        return recordWeek;
    }

    public void setRecordWeek(int recordWeek) {
        this.recordWeek = recordWeek;
    }

    public int getRecordDay() {
        return recordDay;
    }

    public void setRecordDay(int recordDay) {
        this.recordDay = recordDay;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }
}

File: app/src/main/java/com/ywwynm/everythingdone/model/HabitRecord.kt
package com.ywwynm.everythingdone.model

import android.database.Cursor
import androidx.annotation.IntDef
import java.time.Instant
import java.time.ZoneId
import java.time.temporal.WeekFields

/**
 * Created by ywwynm on 2016/2/11.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * model layer for table "habit_records"
 */
open class HabitRecord(
    var id: Long,
    var habitId: Long,
    var habitReminderId: Long,
    var recordTime: Long,
    var recordYear: Int,
    var recordMonth: Int,
    var recordWeek: Int,
    var recordDay: Int,
    var type: Int
) {

    constructor(habitId: Long, habitReminderId: Long) : this(
        0L, habitId, habitReminderId, System.currentTimeMillis(),
        0, 0, 0, 0, TYPE_FINISHED
    ) {
        val dt = Instant.ofEpochMilli(recordTime).atZone(ZoneId.systemDefault())
        recordYear = dt.getYear()
        recordMonth = dt.getMonthValue()
        recordWeek = dt.get(WeekFields.ISO.weekOfWeekBasedYear())
        recordDay = dt.getDayOfMonth()
    }

    constructor(habitId: Long, habitReminderId: Long, recordTime: Long) : this(
        0L, habitId, habitReminderId, recordTime,
        0, 0, 0, 0, TYPE_FINISHED
    ) {
        val dt = Instant.ofEpochMilli(recordTime).atZone(ZoneId.systemDefault())
        recordYear = dt.getYear()
        recordMonth = dt.getMonthValue()
        recordWeek = dt.get(WeekFields.ISO.weekOfWeekBasedYear())
        recordDay = dt.getDayOfMonth()
    }

    constructor(c: Cursor?) : this(
        c!!.getLong(0), c.getLong(1), c.getLong(2), c.getLong(3),
        c.getInt(4), c.getInt(5), c.getInt(6), c.getInt(7), c.getInt(8)
    )

    @IntDef(0, 1, 2, 3)
    @Retention(AnnotationRetention.SOURCE)
    annotation class Type

    companion object {
        const val TYPE_FINISHED: Int = 0
        const val TYPE_CANCEL_FINISHED: Int = 1
        const val TYPE_FAKE_FINISHED: Int = 2
        const val TYPE_FAKE_CANCEL_FINISHED: Int = 3
    }
}
